### ISSUE TITLE
Standardize how we iterate over objects

### ISSUE THREAD

**Issue by markocikos** [liferay/liferay-frontend-guidelines#53]
I've seen all below used in portal, and there are probably others.

1. 
```
for (const key in obj) {
    if (hasOwnProperty(key)) {
        // do stuff
    }
}
```
2. 
```
for (const key in obj) {
    // do stuff, we don't care about fringe inheritance case
}
```
3.
```
Object.keys(obj).forEach(key => {
    // do stuff
});
```
4.
```
for (const key of Object.keys(obj)) {
    // do stuff
}
```
5.
```
for (const [key, value] of Object.entries(obj)) {
    // do stuff
}
```
6.
```
for (let i = 0, l = Object.keys(obj).length; i < l; i++) {
    // do stuff
}
```

Usage of `for..on` is currently flagged as warning in eslint, although (5) is my favorite iteration  method. What should be the recommended method?

**Comment by bryceosterhaus** [comment]
Personally, I tend to lean towards #3. Just what I've grown used to at least.

**Comment by wincent** [comment]
IIRC, the reason we flag (5) is that as a baseline, historically, we preferred to not require the use of a polyfill if we could avoid it. Having said that, I think we do have the necessary polyfills in liferay-portal. Then in our other open source projects, I think both our usage and our recommendations about polyfills vary. As part of this discussion, we should probably deal with whether we want that lint now, and where.

About the options you listed, @markocikos, I don't have strong opinions: 4 and 5 seem "best", but 3 seems fine too.

Another option, which I am not actually suggesting, is to create utilities:

```javascript
const newObject = mapObject(object, (key, value) => {
   return value + 'something else';
});

forEachObject(oblect, (key, value) => {
  // etc
});
```

**Comment by julien** [comment]
Hey guys,

I don't really have a strong opinion, although I'd personally avoid number 2.

I also think that we can't "standardize" everything, in this case I don't see why we couldn't use various ways of iterating through objects. If our rules are too strict it's not going to help either.

**Comment by markocikos** [comment]
That's a good point, @julien 

It is probably best to allow all, except where there is a clear reason why not to use it, such as (2).

In that light, I would lobby that we remove the restriction from `for..on`. We should embrace ES6 patterns where they make sense, and not be afraid of polyfills :)

**Comment by julien** [comment]
@markocikos I agree that we should embrace ES6 and above.

If I remember correctly in the case of `for ... of` we're restricing it's use because it's hard to poyfill in IE (because of `Symbol`) but I might be wrong.

**Comment by oliveiraaraujo** [comment]
I agree with @markocikos because I realize that each type of iterative form has its own use and not just performance or syntax.

**Comment by matuzalemsteles** [comment]
I also agree not to have just one case that we should use but rather the cases that we should not use and why, but personally I also do not have a strong opinion about it.

> If I remember correctly in the case of for ... of we're restricing it's use because it's hard to poyfill in IE (because of Symbol) but I might be wrong.

yes, if you only use babel 7, it was transforming to use symbol, in my view had been a regression and previous versions of babel 7, this did not exist... I do not remember what issue this discussion had, I'll add it here when I find it.

> In that light, I would lobby that we remove the restriction from for..of. We should embrace ES6 patterns where they make sense, and not be afraid of polyfills :)

I'm a bit reluctant to use many polyfills, this increases the load of files we're sending to the client side, this may weigh since they're larger often. But I like how the portal handles this, sending only to IE 🙂.

**Comment by wincent** [comment]
Seeing as I was just chatting to @victorg1991 about this in Slack, wanted to share some links from that chat:

[Here](https://babeljs.io/docs/en/babel-plugin-transform-for-of) you can see what Babel transpiles `for`/`of` to:

```javascript
for (var _iterator = foo[Symbol.iterator](), _step; // etc
```

and [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator) is the MDN pages showing IE lacks support.

Finally, [here is the world of pain](https://www.google.com/search?q=ie+symbol.iterator&oq=ie+sy&aqs=chrome.0.69i59j69i57j0l4.999j0j4&sourceid=chrome&ie=UTF-8) that Google reports when you search for "ie symbol.iterator").

I know we have `for`/`of` uses currently in liferay-portal, I don't the current state of [our IE polyfills](https://github.com/liferay/liferay-portal/blob/056b85730eedc6f417c68469b8f719341d220ee2/modules/apps/frontend-compatibility/frontend-compatibility-ie/build.gradle) with respect to `Symbol.iterator`, and I therefore don't know whether there are places where behavior in IE might be a bit shaky.
